components/fxa-client/src/push.rs

```diff
@@ -0,0 +1,139 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+use crate::{error::*, AccountEvent, FirefoxAccount};
+use serde_derive::Deserialize;
+
+impl FirefoxAccount {
+    /// Handle any incoming push message payload coming from the Firefox Accounts
+    /// servers that has been decrypted and authenticated by the Push crate.
+    ///
+    /// Due to iOS platform restrictions, a push notification must always show UI,
+    /// and therefore we only retrieve 1 command per message.
+    ///
+    /// **💾 This method alters the persisted account state.**
+    pub fn handle_push_message(&mut self, payload: &str) -> Result<Vec<AccountEvent>> {
+        let payload = serde_json::from_str(payload)?;
+        match payload {
+            PushPayload::CommandReceived(CommandReceivedPushPayload { index, .. }) => {
+                if cfg!(target_os = "ios") {
+                    self.fetch_device_command(index)
+                        .map(|cmd| vec![AccountEvent::IncomingDeviceCommand(Box::new(cmd))])
+                } else {
+                    self.poll_device_commands().map(|cmds| {
+                        cmds.into_iter()
+                            .map(|cmd| AccountEvent::IncomingDeviceCommand(Box::new(cmd)))
+                            .collect()
+                    })
+                }
+            }
+            PushPayload::ProfileUpdated => {
+                self.state.last_seen_profile = None;
+                Ok(vec![AccountEvent::ProfileUpdated])
+            }
+            PushPayload::DeviceConnected(DeviceConnectedPushPayload { device_name }) => {
+                Ok(vec![AccountEvent::DeviceConnected { device_name }])
+            }
+            PushPayload::DeviceDisconnected(DeviceDisconnectedPushPayload { device_id }) => {
+                let local_device = self.get_current_device()?;
```

Comment:
Calling `get_current_device` will make a network call to list all the devices. Would it be better to use `get_current_device_id` to retrieve it from local state?